### PR TITLE
enhance: split the containerd daamon control from ctrd

### PR DESCRIPTION
**Pull request by fuweid**
Signed-off-by: Wei Fu <fuweid89@gmail.com>

### Ⅰ. Describe what this PR did

prepare for containerdV1.2

1. make ctrd client package simple
2. add health post check to containerd
3. the default config of containerdV1.2 is different from V1.0.3. The
supervisord package can help us to use configuration file to setup
containerd. It's easy to maintain.

### Ⅱ. Does this pull request fix one issue?

NONE

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

No need

### Ⅳ. Describe how to verify it

When pouchd starts, it should connect to containerd.
When pouchd exits, it should shutdown the containerd.

### CODE REVIEW

File: ctrd/supervisord/daemon.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+		if !strings.HasPrefix(e, "NOTIFY_SOCKET") {
+			cmd.Env = append(cmd.Env, e)
+		}
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// reap the contaienrd process when it has been killed
```
Comment (HusterWan) [inline ctrd/supervisord/daemon.go]: %s/contaienrd/containerd

Comment (fuweid) [reply ctrd/supervisord/daemon.go]: good catch!

File: daemon/daemon.go [hunk 2]
```diff
@@ -66,18 +69,33 @@ func NewDaemon(cfg *config.Config) *Daemon {
...
+	}
+
+	ctrdDaemon, err := supervisord.Start(context.TODO(),
+		filepath.Join(cfg.HomeDir, "containerd/root"),
+		filepath.Join(cfg.HomeDir, "containerd/state"),
+		ctrdDaemonOpts...,
+	)
+	if err != nil {
+		logrus.Errorf("failed to start containerd: %v", err)
+		return nil
```
Comment (HusterWan) [inline daemon/daemon.go]: Why return nil ?

**Comment by HusterWan** [comment]
LGTM

File: daemon/daemon.go [hunk 3]
```diff
@@ -267,15 +286,17 @@ func (d *Daemon) Shutdown() error {
 	}
 
 	logrus.Debugf("Start cleanup containerd...")
+	if err := d.ctrdClient.Cleanup(); err != nil {
+		errMsg = fmt.Sprintf("%s\n", err.Error())
+	}
 
-	if err := d.containerd.Cleanup(); err != nil {
+	if err := d.ctrdDaemon.Stop(); err != nil {
```
Comment (allencloud) [inline daemon/daemon.go]: If `d.ctrdClient.Cleanup()` returns non-nil as well, here we will overwrite the errMsg. I think this has some potential possibility to miss information. How about using string append way to still keep the message? @fuweid

Comment (fuweid) [reply daemon/daemon.go]: Yes. There is some issues about handing error. basically, I will update this part in following PR, not in this pr.

File: ctrd/client_opts.go [hunk 4]
```diff
@@ -3,30 +3,15 @@ package ctrd
 import "fmt"
 
 type clientOpts struct {
-	startDaemon            bool
-	debugLog               bool
 	rpcAddr                string
```
[the remark this answers is not in the document]
Comment (fuweid) [reply ctrd/client_opts.go]: this part will be updated in the following pr. thanks

**Comment by allencloud** [comment]
LGTM and will merge this until the CI passes.

**Pull request merged**